### PR TITLE
[KGE] Add some modeling code

### PR DESCRIPTION
**Pull request by nshah-sc**
Adding some files which support:

- the core heterogeneous graph embedding model which enables various combinations of in-batch vs. random sampling, src/dst operators, negative sampling corruptions, scoring functions etc.
- some negative sampling utilities
- tests for these utilities
- all code is written to be [torch.fx-traceable](https://docs.pytorch.org/docs/stable/fx.html)  -- this requires following certain patterns that avoid [well-known pitfalls](https://docs.pytorch.org/docs/stable/fx.html#limitations-of-symbolic-tracing), including dynamic control flow, and using non-torch functions and following miscellanea carefully.  

Where is the documentation for this feature?: N/A

Did you add automated tests or write a test plan?

***Updated Changelog.md?*** NO

***Ready for code review?:*** YES

### CODE REVIEW

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+
+    Args:
+        model_config (ModelConfig): Configuration object containing model parameters.
+    """
+
+    def __init__(self, model_config: ModelConfig):
+        """Initialize the model with the given embedding configurations."""
+        super().__init__()
+
+        self.num_edge_types = model_config.num_edge_types
```
Comment (svij-sc) [inline python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py:41]: I guess we are trying to learn a unique mapping per edge type?
Is this common ?

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+        old_phase = self._phase
+        self._phase = phase
+        logger.info(f"Changed model phase from {old_phase} to {phase}")
+
+    @property
+    def phase(self) -> ModelPhase:
+        return self._phase
+
+    def fetch_src_and_dst_embeddings(
+        self, edge_batch: EdgeBatch
```
Comment (svij-sc) [inline python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py:123]: where is EdgeBatch defined?

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py (hunk 1 above)
Comment (nshah171) [reply python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py:41]: Pretty common -- e.g. see https://arxiv.org/pdf/2309.12501 Tables 2/3.  These typically all involve learning some parameters per each relation/edge-type r.  the idea is the function of similarity between  src / dst (or head / tail) should be relation-specific.

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/negative_sampling.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+    B, D = src_embeddings.shape
+    N = batch_src_embeddings.shape[0]
+
+    # Precompute similarity matrix between [B] queries and [N] candidates
+    if scoring_function == SimilarityType.DOT:
+        sim_fn = lambda x, y: x @ y.T
+    elif scoring_function == SimilarityType.COSINE:
+        sim_fn = lambda x, y: F.normalize(x, dim=1) @ F.normalize(y, dim=1).T
+    elif scoring_function == SimilarityType.EUCLIDEAN:
```
[the remark this answers is not in the document]
Comment (nshah-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/model/negative_sampling.py:348]: its complete i think -- this if/else block is just assigning the value of sim_fn which is used to compute similarities below.

also what's wrong with ValueError for this as i used below?  its meant to be used when we receive an argument with inappropriate/invalid value IIUC

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+    ModelPhase,
+    SimilarityType,
+)
+
+from gigl.common.logger import Logger
+
+logger = Logger()
+
+
+class HeterogeneousGraphSparseEmbeddingModel(nn.Module):
```
[the remark this answers is not in the document]
Comment (nshah-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py:30]: added the link and another survey paper resource.

re API: i agree the API you shared is probably cleaner.  The main reason I put everything in "one giant model" is because there is work required to make whatever model is used compatible / working with torch.fx.  This requires some of the implementation choices made in the giant model currently and it is easier to build/test that one giant model for being fx-able rather than test a variety of them. 

in practice I think we could actually do something like what you suggested if we just break up the API appropriately, and have these models inherit / reuse fx-able implementations from the ones built now.  

i won't scope for this PR but adding a comment

**Comment by nshah-sc** [comment]
> Also, curious if we had thought to closely reflect the pyg KGE models API here instead of roling out our custom versions?
https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#kge-models

Left a comment  which contains part of the answer: https://github.com/Snapchat/GiGL/pull/293#discussion_r2335258397

What we _could_ do: We could very likely mimic the structure where we have one module per KGE method, and define its `forward` by using some torchrec modules but otherwise ~copying other operations. 

What we _can't_ do:  Even if we do the above, there will be some material changes in the dataloading/negative sampling logic between the two worlds.  in PyG, all the models like `TransR`, `TransE` etc inherit from a `KGEModel` class, which looks [like this](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/kge/base.html#KGEModel).  This class has inbuilt functionalities for loading, loss computation, metric computation, negative sampling, etc. which work on small-scale and are opinionated, but in our case due to added flexibility desiderata and constraints of scale (torchrec, fx-ability) we would likely want to organize them differently (though this may not affect the user-facing API much)

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/negative_sampling.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (svij-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/model/negative_sampling.py:348]: ahhh... man that was confusing lol.
Can we define `def sim_fn(x, y):` somewhere else as `euclidiean` scoring and assign it here?

Alternatively, `scoring_function.impl`.

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/negative_sampling.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    )  # [B, 1]
+
+    # Negative similarity matrix (B x N), depends on corruption side
+    if corrupt_side == NegativeSamplingCorruptionType.SRC:
+        neg_sim_matrix = sim_fn(batch_src_embeddings, dst_embeddings)  # [N, B]
+        neg_sim_matrix = neg_sim_matrix.T  # [B, N]
+    elif corrupt_side == NegativeSamplingCorruptionType.DST:
+        neg_sim_matrix = sim_fn(src_embeddings, batch_dst_embeddings)  # [B, N]
+    elif corrupt_side == NegativeSamplingCorruptionType.BOTH:
+        is_src_corruption = torch.rand_like(
```
[the remark this answers is not in the document]
Comment (svij-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/model/negative_sampling.py:382]: but this isnt deterministic right?
Is there a way to confirm this?

**Review by svij-sc — APPROVED** [review]
Thanks for addressing the comments.

File: python/gigl/experimental/knowledge_graph_embedding/lib/data/edge_batch.py [hunk 6]
```diff
@@ -0,0 +1,362 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from functools import partial
+from typing import Dict, List, Tuple
```
Comment (kmontemayor2-sc) [inline python/gigl/experimental/knowledge_graph_embedding/lib/data/edge_batch.py:5]: fyi I'm trying to migrate to just using `dict, list, tuple`, etc. 

I think our cursor rules should auto generate them.

File: python/gigl/experimental/knowledge_graph_embedding/tests/test_against_batch_contrastive_similarity_relationwise.py
Comment (kmontemayor2-sc) [inline python/gigl/experimental/knowledge_graph_embedding/tests/test_against_batch_contrastive_similarity_relationwise.py:1]: Do these tests run? We should at minimum need an `__init__.py` here right?

If we moved this to `gigl/tests/unit/experimental` or equivalent they should be automatically discovered :)

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+
+    def __init__(self, model_config: ModelConfig):
+        """Initialize the model with the given embedding configurations."""
+        super().__init__()
+
+        self.num_edge_types = model_config.num_edge_types
+        self.node_emb_dim = model_config.node_embedding_dim
+        self.training_sampling_config = model_config.training_sampling
+        self.validation_sampling_config = model_config.validation_sampling
+        self.testing_sampling_config = model_config.testing_sampling
```
Comment (kmontemayor2-sc) [inline python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py:52]: hmmm do all of these need to be accessed later? 

I think it could be simpler to assign `active_sampling_config` directly in init, would reduce indirection here.

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+                self.active_sampling_config.positive_edge_batch_size :
+            ],
+        )
+        pos_labels, batch_neg_labels = (
+            labels[: self.active_sampling_config.positive_edge_batch_size],
+            labels[self.active_sampling_config.positive_edge_batch_size :],
+        )
+
+        pos_logits: torch.Tensor
+        pos_labels: torch.Tensor
```
Comment (kmontemayor2-sc) [inline python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py:295]: btw we should really avoid doing these pre-declarations.

With them, the below will [pass mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=b4e86636cc9e2fc1cef5f71334f2ce8a)

```
a: str
a.startswith("foo")
```

but without them, [mypy will fail](https://mypy-play.net/?mypy=latest&python=3.12&gist=387e36121d4b779af947b42d8c6c7bf9) (this is good, referencing undeclared variables is an error, we should catch it as early as possible.

```
a.startswith("foo")
```

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+        neg_logits = list()
+        neg_labels = list()
```
Comment (kmontemayor2-sc) [inline python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py:297]: nit, can we add types here? 
```suggestion
        neg_logits: list[torch.Tensor] = list()
        neg_labels: list[torch.Tensor] = list()
```

File: python/gigl/experimental/knowledge_graph_embedding/lib/data/edge_batch.py [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+
+    # Determine the number of edges
+    num_edges = len(inputs)
+
+    # Preallocate torch tensors
+    edges = torch.empty((num_edges, 2), dtype=torch.int32)
+    condensed_edge_types = torch.empty(num_edges, dtype=torch.int32)
+
+    # Fill the preallocated torch tensors using direct indexing
+    for i, row in enumerate(inputs):
```
Comment (kmontemayor2-sc) [inline python/gigl/experimental/knowledge_graph_embedding/lib/data/edge_batch.py:287]: how big is `inputs` in practice? And does this code get called in a hot loop? I'm concerned this could end up being slow.

But I guess that's a bridge we can cross when we get to.

**Review by kmontemayor2-sc — COMMENTED** [review]
I think you have the necessary approvals already, left some mostly nits.

Would prefer if we move the tests folder to somewhere they get run though :)

File: python/gigl/experimental/knowledge_graph_embedding/lib/data/edge_batch.py (hunk 10 above)
Comment (nshah-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/data/edge_batch.py:287]: this does run in a hot loop although this has not proven to be a major performance bottleneck in my observation.  there may be future optimizations which are warranted here but not high-prio at the time

File: python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py (hunk 7 above)
Comment (nshah-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/model/heterogeneous_graph_model.py:52]: yeah, they do all need to be accessed later.   

i'll leave this here for now as this class may be worth a bigger refactor /decomposition later

File: python/gigl/experimental/knowledge_graph_embedding/lib/data/edge_batch.py (hunk 6 above)
Comment (nshah-sc) [reply python/gigl/experimental/knowledge_graph_embedding/lib/data/edge_batch.py]: done
